README.md

```diff
@@ -12,63 +12,159 @@ purposes:
+$ cp api_keys.conf.example /tmp/api_keys.conf
 ```
 
-## Local development
+Set the API keys appropriately:
+
+- Generate an API key to be given to Echoes during the GitLab integration installation
 
 ```console
-# navigate to the proxy folder
-cd nginx-proxy
+$ openssl rand -base64 18
+oLAVcK2LAzfMpYXT10ymK1qL
+```
+
+- Generate a GitLab [Personal Token](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#create-a-personal-access-token)
+
+The final content should resemble to the example below:
+
+```txt
+map $http_PRIVATE_TOKEN $gitlab_token {
+    default "";
+
+    "oLAVcK2LAzfMpYXT10ymK1qL" "glpat-ZjNEWE6JPQ6YX5YBEfYz";
+}
+```
 
-# build the proxy
-docker build . -t gitlab-proxy
+### Step 5: Run the proxy
 
-# create a .env file
-cp .env.example .env
+The proxy expects the file `api_keys.conf` to me mounted to `/etc/nginx/api_keys`.
 
-# run the proxy
+With the command below the proxy will run on the port `8080`.
+
+```console
 docker run \
     --env-file .env \
-    -v /tmp/api_keys.conf:/etc/nginx/api_keys.conf:ro \
+    -v /tmp/api_keys.conf:/etc/nginx/api_keys/api_keys.conf:ro \
     -p 8080:80 gitlab-proxy
 ```
 
+### Step 6: Call the GitLab API
+
+Set the HTTP Header `PRIVATE-TOKEN` to the APi key set in the `api_keys.conf` file earlier, e,g `oLAVcK2LAzfMpYXT10ymK1qL`
+
 ```console
-# call the GitLab API via the proxy
 curl --location --request GET 'http://0.0.0.0:8080/api/v4/groups/<REPLACE_ME>/members' \
     --header 'Content-Type: application/json' \
-    --header 'PRIVATE-TOKEN: <REPLACE_ME>'
+    --header 'PRIVATE-TOKEN: oLAVcK2LAzfMpYXT10ymK1qL'
+```
+
+## Configuration with an upstream directive (Optional)
+
+For on-premises `GitLab` instances, an upstream configuration could be necessary.
+
+Below are the steps in order to use an upstream configuration with the proxy.
+
+### Step 1: Copy the `api_backends.conf` file
+
+Copy the `api_backends.conf` file containing an example of upstream into a mount location on the host.
+
+```console
+$ cp api_backends.conf /tmp/api_backends.conf
+```
+
+### Step 2: Configure the upstream
+
+Modify the copied `api_backends.conf` file in order to map the gitlab instance IP(s) of your choice.
+
+### Step 3: Set the `GITLAB_URL` environment variable
+
+Set the environment variable `GITLAB_URL` from the `.env` file as follow: `<protocol><upstream_name>` e,g `http://example`
+
+```txt
+GITLAB_URL=http://example
+```
+
+### Step 4: Run the proxy
+
+The proxy expects the file `api_backends.conf` to me mounted to `/etc/nginx/api_backends/api_backends.conf`.
+
+```console
+docker run \
+    --env-file .env \
+    -v /tmp/api_keys.conf:/etc/nginx/api_keys/api_keys.conf:ro \
+    -v /tmp/api_backends.conf:/etc/nginx/api_backends/api_backends.conf:ro \
+    -p 8080:80 gitlab-proxy
 ```
+
+## Restricted access
+
+It is highly advised to restrict the access to the proxy to only known Echoes IPs.
+To enable this restriction uncomment the related part in the `gitlab-proxy.conf` file.
```

Comment:
I'm curious why not make this the default? We could keep this documentation section to say: "The proxy is preconfigured to restrict incoming connections to Echoes exit IP addresses:"

Reply:
Discussed on Discord and agreed that we should make it a default.